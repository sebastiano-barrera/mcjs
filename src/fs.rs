use std::{
    collections::HashMap,
    path::{Path, PathBuf},
};

use crate::bytecode::{self, ModuleId};
use crate::bytecode_compiler::Loader;

pub struct FileLoader {
    module_paths: Vec<PathBuf>,
    module_ids: HashMap<PathBuf, bytecode::ModuleId>,
    path_of_module: HashMap<bytecode::ModuleId, PathBuf>,
    next_module_id: u16,
}

impl FileLoader {
    pub fn new(module_paths: Vec<PathBuf>) -> Self {
        let module_paths = module_paths
            .into_iter()
            .map(|p| p.canonicalize().expect("could not canonicalize path"))
            .collect();
        FileLoader {
            module_paths,
            module_ids: HashMap::new(),
            path_of_module: HashMap::new(),
            next_module_id: 0,
        }
    }
}

impl Loader for FileLoader {
    fn get_module_id(&mut self, relative_path: &str) -> Option<bytecode::ModuleId> {
        let relative_path = Path::new(relative_path);

        let file_path = self.module_paths.iter().find_map(|mp| {
            let candidate_file_path = mp.join(&relative_path);
            match candidate_file_path.is_file() {
                true => Some(candidate_file_path),
                false => None,
            }
        })?;

        let mid = if let Some(mid) = self.module_ids.get(&file_path) {
            *mid
        } else {
            let mid = bytecode::ModuleId(self.next_module_id);
            self.module_ids.insert(file_path.clone(), mid);
            self.path_of_module.insert(mid, file_path);
            self.next_module_id += 1;
            mid
        };

        Some(mid)
    }

    fn read_source(&self, module_id: crate::bytecode::ModuleId) -> String {
        let path = self.path_of_module.get(&module_id).unwrap();
        std::fs::read_to_string(path).expect("read error")
    }
}

pub struct MockLoader {
    mid_of_file: HashMap<String, bytecode::ModuleId>,
    code_of_module: HashMap<bytecode::ModuleId, String>,
}

impl MockLoader {
    pub fn new() -> Self {
        MockLoader {
            mid_of_file: HashMap::new(),
            code_of_module: HashMap::new(),
        }
    }
    pub fn add_module(&mut self, filename: String, module_id: bytecode::ModuleId, code: String) {
        self.mid_of_file.insert(filename, module_id);
        self.code_of_module.insert(module_id, code);
    }
}

impl Loader for MockLoader {
    fn get_module_id(&mut self, filename: &str) -> Option<bytecode::ModuleId> {
        self.mid_of_file.get(filename).copied()
    }

    fn read_source(&self, module_id: bytecode::ModuleId) -> String {
        self.code_of_module.get(&module_id).cloned().unwrap()
    }
}

pub struct CombinedLoader {
    inner: Vec<Box<dyn Loader>>,
    module_ids: HashMap<bytecode::ModuleId, (usize, bytecode::ModuleId)>,
    last_module_id: u16,
}

impl CombinedLoader {
    pub fn new(inner: Vec<Box<dyn Loader>>) -> Self {
        CombinedLoader {
            inner,
            module_ids: HashMap::new(),
            last_module_id: 0,
        }
    }

    fn next_module_id(&mut self) -> ModuleId {
        let new_module_id = bytecode::ModuleId(self.last_module_id);
        self.last_module_id += 1;
        new_module_id
    }
}

impl Loader for CombinedLoader {
    fn get_module_id(&mut self, filename: &str) -> Option<bytecode::ModuleId> {
        for (ndx, loader) in self.inner.iter_mut().enumerate() {
            if let Some(inner_module_id) = loader.get_module_id(filename) {
                let new_module_id = self.next_module_id();
                self.module_ids
                    .insert(new_module_id, (ndx, inner_module_id));
                return Some(new_module_id);
            }
        }

        None
    }

    fn read_source(&self, module_id: bytecode::ModuleId) -> String {
        let (loader_ndx, inner_mod_id) = *self.module_ids.get(&module_id).unwrap();
        let loader = &self.inner[loader_ndx];
        loader.read_source(inner_mod_id)
    }
}
