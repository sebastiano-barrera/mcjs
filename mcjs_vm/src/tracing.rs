#![cfg_attr(not(test), allow(unused_variables))]

pub fn section(header: &str) -> Section {
    internal::section(header)
}
pub type Section = internal::Section;

#[cfg(all(not(test), not(feature = "tracing")))]
mod internal {
    use std::fmt::Debug;

    pub fn section(_header: &str) -> Section {
        Section
    }
    pub struct Section;

    impl Section {
        pub fn log_value<T: Debug>(&self, _tag: &str, _value: &T) {}
        pub fn log(&self, _tag: &str, _value: &str) {}
    }
}

#[cfg(any(test, feature = "tracing"))]
mod internal {
    use lazy_static::lazy_static;

    use std::cell::RefCell;
    use std::fmt::Debug;
    use std::fs::File;
    use std::io::Write;
    use std::path::PathBuf;
    use std::sync::{Arc, Mutex, MutexGuard};
    use std::thread_local;

    //
    // Public API
    // (does not require a handle, simply forwards into *the* thread-local handle)
    //

    #[must_use]
    pub fn section(header: &str) -> Section {
        THREAD_LOGGER.with(|logger| {
            let mut logger = logger.borrow_mut();
            logger.start_section(header);
        });

        Section
    }

    pub struct Section;
    impl Section {
        pub fn log_value<T: Debug>(&self, tag: &str, value: &T) {
            THREAD_LOGGER.with(|logger| {
                let mut logger = logger.borrow_mut();
                logger.log_value(tag, value);
            });
        }

        pub fn log(&self, tag: &str, value: &str) {
            THREAD_LOGGER.with(|logger| {
                let mut logger = logger.borrow_mut();
                logger.log(tag, value);
            })
        }
    }

    impl Drop for Section {
        fn drop(&mut self) {
            THREAD_LOGGER.with(|logger| {
                let mut logger = logger.borrow_mut();
                logger.end_section();
            })
        }
    }

    struct SinkConfig {
        sink_filename: PathBuf,
    }
    impl Default for SinkConfig {
        fn default() -> Self {
            SinkConfig {
                sink_filename: PathBuf::from("/dev/null"),
            }
        }
    }
    impl SinkConfig {
        fn from_env() -> Self {
            let sink_filename: PathBuf = std::env::var("MCJS_TRACING_SINK")
                .unwrap_or_else(|_| "/dev/null".to_string())
                .into();
            SinkConfig { sink_filename }
        }
    }

    //
    // Thread-local part
    //

    thread_local! {
        static THREAD_LOGGER: RefCell<Logger> = RefCell::new(Logger::new());
    }

    struct Logger {
        thread_prefix: String,
        buf: String,
        sections_depth: usize,
    }

    impl Default for Logger {
        fn default() -> Self {
            let thread_id = format!("{:?}: ", std::thread::current().id());
            Logger {
                thread_prefix: thread_id,
                buf: String::new(),
                sections_depth: 0,
            }
        }
    }

    impl Logger {
        fn new() -> Self {
            Logger::default()
        }

        fn start_section(&mut self, header: &str) {
            self.writeln(&format!("<{}>", header));
            self.sections_depth += 1;
        }

        fn log_value<T: Debug>(&mut self, tag: &str, value: &T) {
            self.writeln(&format!("{}: {:#?}", tag, value));
        }
        fn log(&mut self, tag: &str, value: &str) {
            self.writeln(&format!("{}: {}", tag, value));
        }

        fn end_section(&mut self) {
            assert!(self.sections_depth > 0);
            self.sections_depth -= 1;
            self.writeln("</>");
        }

        fn writeln(&mut self, text: &str) {
            use std::fmt::Write;

            self.buf.clear();
            for line in text.split('\n') {
                self.buf.push_str(&self.thread_prefix);
                for _ in 0..self.sections_depth {
                    write!(self.buf, "    ").unwrap();
                }

                write!(self.buf, "{}\n", line).unwrap();
            }

            // flush global sink file. inefficient, but tracing is often employed to debug VM
            // crashes. we might be crashing right after returning, so let's work to make it more
            // likely that the data is not lost.
            let mut sink = get_global_sink();
            sink.write(&self.buf);
        }
    }

    //
    // Global part (process-wide, thread-nonlocal)
    //
    lazy_static! {
        static ref GLOBAL_SINK: Arc<Mutex<Sink>> =
            Arc::new(Mutex::new(Sink::from_config(SinkConfig::from_env())));
    }

    fn get_global_sink() -> MutexGuard<'static, Sink> {
        GLOBAL_SINK.lock().expect("could not lock global logger!")
    }

    #[cfg(test)]
    fn set_global_sink_memory() {
        let mut gsink = GLOBAL_SINK.lock().expect("could not lock global logger!");
        *gsink = Sink::Memory(String::new());
    }

    #[cfg(test)]
    fn read_memory_buffer_this_thread() -> String {
        let mut gsink = GLOBAL_SINK.lock().expect("could not lock global logger!");
        let buf = match &mut *gsink {
            Sink::Memory(buf) => buf,
            _ => panic!("global sink not initialized as memory buffer"),
        };

        let prefix = format!("{:?}: ", std::thread::current().id());
        let mut tl_buf = String::new();
        for line in buf.lines() {
            if let Some(tl_line) = line.strip_prefix(&prefix) {
                tl_buf.push_str(tl_line);
                tl_buf.push('\n');
            }
        }

        tl_buf
    }

    enum Sink {
        File(File),
        Memory(String),
    }
    impl Sink {
        fn from_config(config: SinkConfig) -> Self {
            let file = File::create(config.sink_filename).expect("could not open logger sink file");
            Sink::File(file)
        }

        fn write(&mut self, data: &str) {
            match self {
                Sink::File(f) => {
                    f.write(data.as_bytes()).unwrap();

                    // inefficient, but tracing is often employed when the program is crashing. we might be
                    // crashing right after returning from this call, so let's flush immediately.
                    f.flush().unwrap();
                }
                Sink::Memory(buf) => {
                    // no flushing necessary
                    buf.push_str(data);
                }
            }
        }
    }

    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        fn test_section() {
            set_global_sink_memory();

            {
                let _s = section("section name");
                _s.log_value("the_value", &(1, 2, 3));
            }

            let output = read_memory_buffer_this_thread();
            // Remove thread ID
            insta::assert_snapshot!(output);
        }
    }
}
