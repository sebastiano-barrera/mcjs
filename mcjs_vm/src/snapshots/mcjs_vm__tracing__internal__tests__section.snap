---
source: mcjs_vm/src/tracing.rs
expression: output
---
<section name>
    the_value: (
        1,
        2,
        3,
    )
    

